base/c.jl

```diff
@@ -503,3 +503,140 @@ end
+    if !Meta.isexpr(call, :call)
+        throw(ArgumentError("@ccall has to take a function call"))
+    end
+
+    # get the function symbols
+    func = let f = call.args[1]
+        f isa Expr ? :(($(f.args[2]), $(f.args[1]))) : QuoteNode(f)
+    end
+
+    # detect varargs
+    varargs = nothing
+    argstart = 2
+    callargs = call.args
+    if length(callargs) >= 2 && Meta.isexpr(callargs[2], :parameters)
+        argstart = 3
+        varargs = callargs[2].args
+    end
+
+    # collect args and types
+    args = []
+    types = []
+
+    function pusharg!(arg)
+        if !Meta.isexpr(arg, :(::))
+            throw(ArgumentError("args in @ccall need type annotations. '$(repr(arg))' doesn't have one."))
+        end
+        push!(args, arg.args[1])
+        push!(types, arg.args[2])
+    end
+
+    for i in argstart:length(callargs)
+        pusharg!(callargs[i])
+    end
+    # add any varargs if necessary
+    nreq = 0
+    if !isnothing(varargs)
+        nreq = length(args)
+        for a in varargs
+            pusharg!(a)
+        end
+    end
+
+    return func, rettype, types, args, nreq
+end
+
+function lower(convention, func, rettype, types, args, nreq)
+    lowering = []
+    realargs = []
+    gcroots = []
+    for (i, (arg, type)) in enumerate(zip(args, types))
+        sym = Symbol(string("arg", i, "root"))
+        sym2 = Symbol(string("arg", i, ))
+        earg, etype = esc(arg), esc(type)
+        push!(lowering, :($sym = Base.cconvert($etype, $earg)))
+        push!(lowering, :($sym2 = Base.unsafe_convert($etype, $sym)))
+        push!(realargs, sym2)
+        push!(gcroots, sym)
+    end
+    etypes = Expr(:call, Expr(:core, :svec), types...)
+    exp = Expr(:foreigncall,
+               esc(func),
+               esc(rettype),
+               esc(etypes),
+               nreq,
+               QuoteNode(convention),
+               realargs..., gcroots...)
+    push!(lowering, exp)
+
+    return Expr(:block, lowering...)
+end
+
+
+"""
+    @ccall some_c_function(arg::Type [...])::ReturnType
+
+    @ccall calling_convetion some_c_function(arg::Type)::ReturnType
+
+convert a julia-style function definition to a ccall:
+
+    @ccall link(source::Cstring, dest::Cstring)::Cint
+
+same as:
+
+    ccall(:link, Cint, (Cstring, Cstring), source, dest)
+
+All arguments must have type annotations and the return type must also
+be annotated.
+
+varargs are supported with the following convention:
+
+    @ccall printf("%s = %d"::Cstring ; "foo"::Cstring, foo::Cint)::Cint
+
+Mind the semicolon.
+
+Using functions from other libraries is supported by prefixing
+the function name with the name of the C library, like this:
+
+    const glib = "libglib-2.0"
+    @ccall glib.g_uri_escape_string(uri::Cstring, ":/"::Cstring, true::Cint)::Cstring
+
+The string literal could also be used directly before the symbol of
+the function name, if desired `"libglib-2.0".g_uri_escape_string(...`
+"""
+macro ccall(convention, expr)
```

Comment:
We still need to decide on the placement of the calling convention argument (and other function attributes).

Unless someone else speaks up with a desired API for this (or if it's been already discussed and decided on elsewhere), I'd be inclined to conservatively remove this part from the current PR, and simply default to always using `:ccall`. In the future some syntax for the calling convention and other function attributes can then be added in a backward compatible way. (What I'm worried about is baking in an insufficiently-general syntax for function attributes, and would prefer that design discussion to not further hold up this PR.)

Reply:
At some earlier points it the discussion, it was suggested that the macro should be a drop-in replacement for `ccall`, which allows the specification of a calling convention.

https://github.com/JuliaLang/julia/pull/32748#pullrequestreview-282370657
...
Oh, it was you. XD

Reply:
Right, but the issue of extra function attributes was also brought up in the conversation by @vtjnash, who linked to https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-pure-function-attribute as an example of function attributes in gcc. Note that some of these attributes take arguments so I'm not sure what is desired as a fully general syntax here. Do we want key value pairs for example? Is it ok to have just a calling convention key like `stdcall` without a value? Should it go before or after the actual call expression, etc etc.

If someone has a fully formed idea of how all this should work, or is confident that the current placement of the calling convention is what we want please chime in! (If not, I was just thinking we could get this PR merged without a big diversion into that part of the design space.)

Reply:
understood.

Reply:
Perfect!